packages/paste-core/components/badge/src/utils.ts

```diff
@@ -0,0 +1,66 @@
+import type {TextColorOptions, BackgroundColorOptions} from '@twilio-paste/style-props';
+
+import type {BadgeVariants, BadgeProps} from './types';
+
+export const hasValidButtonVariantProps = ({as, onClick}: Partial<BadgeProps>): boolean => {
+  const clickHandlerIsValid = typeof onClick == 'function';
+  const asPropIsValid = as === 'button';
+  return clickHandlerIsValid && asPropIsValid;
+};
+export const hasValidAnchorVariantProps = ({as, href}: Partial<BadgeProps>): boolean => {
+  const asPropIsValid = as === 'a';
+  const hrefIsValid = typeof href === 'string';
+  return asPropIsValid && hrefIsValid;
+};
+
+type Props = Record<string, any>;
+export const safelySpreadProps = (props: Props, denylist: string[]): Partial<Props> =>
```

Comment:
This is what prompted my comment on safely spread. Ideally we only have one version and it lives in Box

Reply:
Yeah that makes a lot of sense. I really like the idea too of leveraging hooks for side effects, event handling, and state management (internal and interaction). It would not only help our development, but also empower customers to build their own custom components (similar to how we are using Reakit).